stdlib/public/core/Arrays.swift.gyb

```diff
@@ -2177,61 +2177,56 @@ extension _ArrayBufferProtocol {
 }
 
 % for (Self, a_Self) in arrayTypes:
-// NOTE: The '==' and '!=' below only handles array types
-// that are the same, e.g. Array<Int> and Array<Int>, not
-// ArraySlice<Int> and Array<Int>.
 
-/// Returns `true` if these arrays contain the same elements.
-@_inlineable
-public func == <Element : Equatable>(
-  lhs: ${Self}<Element>, rhs: ${Self}<Element>
-) -> Bool {
-  let lhsCount = lhs.count
-  if lhsCount != rhs.count {
-    return false
-  }
+extension ${Self} where Element : Equatable {
+  /// Returns `true` if these arrays contain the same elements.
+  @_inlineable
+  public static func ==(lhs: ${Self}<Element>, rhs: ${Self}<Element>) -> Bool {
+    let lhsCount = lhs.count
+    if lhsCount != rhs.count {
+      return false
+    }
 
-  // Test referential equality.
-  if lhsCount == 0 || lhs._buffer.identity == rhs._buffer.identity {
-    return true
-  }
+    // Test referential equality.
+    if lhsCount == 0 || lhs._buffer.identity == rhs._buffer.identity {
+      return true
+    }
 
-%if Self == 'ArraySlice':
+  %if Self == 'ArraySlice':
 
-  var streamLHS = lhs.makeIterator()
-  var streamRHS = rhs.makeIterator()
+    var streamLHS = lhs.makeIterator()
+    var streamRHS = rhs.makeIterator()
```

Comment:
Anyone know why we're still using iterators to compare slices? It must be strictly more overhead than indexing. cc @airspeedswift

Reply:
I have another PR that gave them custom iterators, which had a speed benefit but was causing some compiler issues. If I revisit that, iterators could be faster.